src/main/java/com/gluonhq/substrate/util/FileOps.java

```diff
@@ -365,4 +373,127 @@ public static void extractFilesFromJar(String extension, Path sourceJar, Path ta
+     * @throws IOException
+     */
+    public static void downloadFile(URL fileUrl, Path filePath) throws IOException {
+        Objects.requireNonNull(fileUrl);
+        Objects.requireNonNull(filePath);
+        try (ReadableByteChannel readableByteChannel = Channels.newChannel(fileUrl.openStream());
+             FileOutputStream fileOutputStream = new FileOutputStream(filePath.toFile());
+             FileChannel fileChannel = fileOutputStream.getChannel()) {
+            fileChannel.transferFrom(readableByteChannel, 0, Long.MAX_VALUE);
+        } catch (IOException e) {
+            throw new IOException("Error downloading from " + fileUrl + "into " + filePath + ": " + e.getMessage() + ", " + Arrays.toString(e.getSuppressed()));
+        }
+    }
+
+    /**
+     * Extracts the files from a given zip file into a target folder, and returns a map
+     * with the names of the files and their checksum values.
+     * @param sourceZip the path of a non null zip file
+     * @param targetDir the path of a folder where the zip file will be extracted
+     * @return a map with the file names and their checksum values
+     * @throws IOException
+     */
+    public static Map<String, String> unzipFile(Path sourceZip, Path targetDir) throws IOException {
+        Objects.requireNonNull(sourceZip);
+        Objects.requireNonNull(targetDir);
+        if (!Files.exists(sourceZip)) {
+            throw new IOException("Error: " + sourceZip + " does not exist");
+        }
+        if (!sourceZip.toString().endsWith(".zip")) {
+            throw new IOException("Error: " + sourceZip + " is not a valid zip file");
+        }
+        if (Files.isRegularFile(targetDir)) {
+            throw new IOException("Error: " + targetDir + " is not a directory");
+        }
+        if (!Files.exists(targetDir)) {
+            Files.createDirectories(targetDir);
+        }
+        Map<String, String> hashes = new HashMap<>();
+        try (ZipInputStream zis = new ZipInputStream(new FileInputStream(sourceZip.toFile()))) {
+            ZipEntry zipEntry = zis.getNextEntry();
+            while (zipEntry != null) {
+                File destFile = new File(targetDir.toFile(), zipEntry.getName());
+                if (zipEntry.isDirectory()) {
+                    if (!destFile.exists()) {
+                        Files.createDirectories(destFile.toPath());
+                    }
+                } else {
+                    byte[] buffer = new byte[1024];
+                    try (FileOutputStream fos = new FileOutputStream(destFile)) {
+                        int len;
+                        while ((len = zis.read(buffer)) > 0) {
+                            fos.write(buffer, 0, len);
+                        }
+                    }
+                    hashes.put(destFile.getName(), calculateCheckSum(destFile));
+                }
+                zipEntry = zis.getNextEntry();
+            }
+            zis.closeEntry();
```

Comment:
Closing may never occur if exception is thrown. `Try/finally` or `Try with resources` is needed.

Reply:
Not sure if I understand it correctly, but the whole zip process is within a try with resources.

Reply:
I meant `zis.closeEntry()` - it is a separate create/close cycle within the overall zip processing.

Reply:
I've realized I wasn't closing each entry. I've moved that to the `while`. The external try with resources will manage to call `zis.close()`.
I still don't see the need to include a try-catch for `zis.closeEntry()`.

Reply:
I might be wrong, but `zis.closeEntry` is still perceptible to failure. Cannot say for sure what the effect of not closing the entry is - might be a memory leak. If you feel good about it - no need to fix